In [2]:
SELECT
    YEAR(OrderDate)     as [Year],
    MONTH(OrderDate)    as [Month],
    COUNT(*)            as [Orders],
    SUM(SubTotal)       as [SubTotal]
FROM Orders
where YEAR(OrderDate) = 2012
GROUP BY YEAR(OrderDate), MONTH(OrderDate)
ORDER BY 1,2

(12 rows affected)

Total execution time: 00:00:00.044

Year Month Orders SubTotal 2012 1 336 3970627,279 2012 2 219 1475426,91 2012 3 304 2975748,2385 2012 4 269 1634600,7984 2012 5 293 3074602,8138 2012 6 390 4099354,3572 2012 7 385 3417953,8699 2012 8 285 2175637,218 2012 9 352 3454151,9405 2012 10 321 2544091,1056 2012 11 383 1872701,9762 2012 12 378 2829404,8189

In [15]:
SELECT 
    *,
    COUNT(*)        OVER(), --aggregáló ablakozás esetében nem kötelezó az order by
    SUM(Orders)     OVER(partition by Year), --összesíti a rendelést évenként
    SUM(Orders)     OVER(partition by Year order by Month), --göngyölt összeg
    SUM(SubTotal)   OVER(partition by Year order by Month), --göngyölt összeg
    MAX(SubTotal)   OVER(partition by Year order by Month rows between unbounded preceding and current row)


FROM (
    SELECT
        YEAR(OrderDate)     as [Year],
        MONTH(OrderDate)    as [Month],
        COUNT(*)            as [Orders],
        SUM(SubTotal)       as [SubTotal]
    FROM Orders
    where YEAR(OrderDate) in (2012, 2011, 2013)
    GROUP BY YEAR(OrderDate), MONTH(OrderDate)
    --ORDER BY 1,2 Subqueryben nem lehet order by
) as s
ORDER BY 1, 2

(32 rows affected)

Total execution time: 00:00:00.230

Year Month Orders SubTotal (No column name) (No column name) (No column name) (No column name) 2011 5 43 503805,9169 32 1607 43 503805,9169 2011 6 141 458910,8248 32 1607 184 962716,7417 2011 7 231 2044600,0034 32 1607 415 3007316,7451 2011 8 250 2495816,7335 32 1607 665 5503133,4786 2011 9 157 502073,8458 32 1607 822 6005207,3244 2011 10 327 4588761,816 32 1607 1149 10593969,1404 2011 11 230 737839,8214 32 1607 1379 11331808,9618 2011 12 228 1309863,2511 32 1607 1607 12641672,2129 2012 1 336 3970627,279 32 3915 336 3970627,279 2012 2 219 1475426,91 32 3915 555 5446054,189 2012 3 304 2975748,2385 32 3915 859 8421802,4275 2012 4 269 1634600,7984 32 3915 1128 10056403,2259 2012 5 293 3074602,8138 32 3915 1421 13131006,0397 2012 6 390 4099354,3572 32 3915 1811 17230360,3969 2012 7 385 3417953,8699 32 3915 2196 20648314,2668 2012 8 285 2175637,218 32 3915 2481 22823951,4848 2012 9 352 3454151,9405 32 3915 2833 26278103,4253 2012 10 321 2544091,1056 32 3915 3154 28822194,5309 2012 11 383 1872701,9762 32 3915 3537 30694896,5071 2012 12 378 2829404,8189 32 3915 3915 33524301,326 2013 1 400 2087872,4623 32 14182 400 2087872,4623 2013 2 325 2316922,1516 32 14182 725 4404794,6139 2013 3 441 3412068,9676 32 14182 1166 7816863,5815 2013 4 428 2532265,9123 32 14182 1594 10349129,4938 2013 5 428 3245623,7551 32 14182 2022 13594753,2489 2013 6 719 5081069,1319 32 14182 2741 18675822,3808 2013 7 1740 4896353,7386 32 14182 4481 23572176,1194 2013 8 1789 3333964,0677 32 14182 6270 26906140,1871 2013 9 1791 4532908,7054 32 14182 8061 31439048,8925 2013 10 1968 4795813,2898 32 14182 10029 36234862,1823 2013 11 2103 3312130,2457 32 14182 12132 39546992,428 2013 12 2050 4075486,6257 32 14182 14182 43622479,0537

Az 5 sor hogyan is működik. Amikor elindul a lekérdezés ugye 1 sor van, ennek az összege 43 mert az ablak 1 soros

a következő iterációban az ablak két soros, így a 1 és 2 összege (43+141) lesz az eredménye.

Ha kiveszed a patrition by-t akkor az ablak az egész lekérdezésre fog vonatkozní, és a göngyölítés a legérdezés végéig <span style="color: var(--vscode-foreground);">megy.</span>

<span style="color: var(--vscode-foreground);">7 Sor</span>

<span style="color: var(--vscode-foreground);">Itt az van definiálva, hogy az ablak folyamatosan nőljön, amíg az év nem változik. Itt lehet olyan, hogy a max érték változik emiatt.</span>

<span style="color: var(--vscode-foreground);">Ha azt szeretnénk, hogy az egész patrition by által megadott ablakban keresse a max-ot, akkor nem kell order by, vagy&nbsp;</span> 

<span style="color: var(--vscode-foreground);">... and unbounded following</span>

<span style="color: var(--vscode-foreground);"><br></span>